src/Rings/mpoly.jl

```diff
@@ -546,11 +551,61 @@ function singular_ring(Rx::MPolyRing{T}, ord::Symbol) where {T <: RingElem}
               cached = false)[1]
 end
 
+function singular(o::Orderings.GenOrdering)
+  v = o.vars
+  @assert minimum(v)+length(v) == maximum(v)+1
+  if o.ord == :lex
+    return Singular.ordering_lp(length(v))
+  elseif o.ord == :degrevlex
+    return Singular.ordering_dp(length(v))
+  elseif o.ord == :weights
+    return Singular.ordering_M(o.wgt)
+  else
+    error("not done yet")
+  end
+end
+
+function singular_ring(Rx::MPolyRing{T}, ord::Orderings.AbsOrdering) where {T <: RingElem}
+  #test if it can be mapped directly to singular:
+  # - consecutive, non-overlapping variables
+  # - covering everything
+  # if this fails, create a matrix...
+  f = Orderings.flat(ord)
+  st = 1
+  iseasy = true
+  for i = 1:length(f)
+    mi = minimum(f[i].vars)
+    ma = maximum(f[i].vars)
+    if mi == st && length(f[i].vars) + st == ma+1
+      st = ma+1
+    else
+      iseasy = false
+      break
+    end
+  end
+
+  if iseasy
+    o = singular(f[1])
+    for i=2:length(f)
+      o = o*singular(f[i])
+    end
+  else
+    o = Singular.ordering_M(Orderings.simplify_weight_matrix(ord))
+  end
+
+  return Singular.PolynomialRing(singular_ring(base_ring(Rx)),
+              [string(x) for x = Nemo.symbols(Rx)],
+              ordering = o,
+              cached = false)[1]
+end
+
+
 #catch all for generic nemo rings
 function Oscar.singular_ring(F::AbstractAlgebra.Ring)
   return Singular.CoefficientRing(F)
 end
 
+#??? needs to coerce into b? assert parent?
 function (b::AbstractAlgebra.Ring)(a::Singular.n_unknown)
   Singular.libSingular.julia(Singular.libSingular.cast_number_to_void(a.ptr))
 end
```

Comment:
Is this function signature a bit too general?
It is very prone to errors as well because of the lack of a parent check. Maybe
```julia
function (b::AbstractAlgebra.Ring)(a::Singular.n_unknown)
  b(Singular.libSingular.julia(Singular.libSingular.cast_number_to_void(a.ptr)))
end
```
If this function signature is not too general, then https://github.com/oscar-system/Singular.jl/blob/master/src/MessyHacks.jl can probably go away.